### ISSUE TITLE
Add support for remote debugger

### ISSUE THREAD

**Issue by dlsniper** [go-lang-plugin-org/go-lang-idea-plugin#620]
Once #588, it would be nice if the debugger could support existing processes and processes on remote hosts.

This ticket is to keep track of that.

**Comment by dlsniper** [comment]
When we have Delve support it would be (somewhat) easy to add.

**Comment by dlsniper** [comment]
I'll close this is as it's related to another ticket, supporting gdb which well, is "deprecated" by delve. Delve will have native remote debugging options as it's running over websockets, see https://github.com/derekparker/delve/pull/88

**Comment by dlsniper** [comment]
I guess it would be nice to reopen this now that we have delve and it's a requested feature, see https://groups.google.com/forum/m/#!topic/golang-nuts/tJys0_JdZDA

**Comment by hiromaily** [comment]
I'm interested in this topic.
Does this project depend on features of delve?

**Comment by dlsniper** [comment]
No, delve support is already available in the plugin, it's just a matter of priorities for this to make its way into the supposed features list.

**Comment by krise3k** [comment]
+1

**Comment by QianJin2013** [comment]
+1

**Comment by Scratch-net** [comment]
Yep, it's a blocker feature for many projects

**Comment by ovjy** [comment]
any idea when this feature will be ready?

**Comment by Yarel-Indegy** [comment]
+1

**Comment by ignatov** [comment]
Hi all,
Does anybody know a way how to understand to which processes Delve can connect to?

**Comment by dlsniper** [comment]
@ignatov Delve needs to be started on the remote server then the API can be contacted like it would be the localhost. Right now we are using ` localhost:port` but we could just as well have a ` remoteHost:port `. Does this help?

**Comment by dlsniper** [comment]
The other option is: ` dlv connect host:port --listen=localhost:somePort ` where ` host:port ` is the remote running delve and ` localhost:somePort ` is what we have today.

**Comment by ignatov** [comment]
I'd like to have a fast way to determine is it possible to connect to the desired process with Delve or not.

**Comment by dlsniper** [comment]
Maybe I've read this wrong, by connecting to a process you mean attach to a running process? Or connect to a remote host?

**Comment by ignatov** [comment]
> attach to a running process
right

**Comment by dlsniper** [comment]
I see, I confused it with the title of the issue which was about connecting to the remote host, sorry.

Maybe doing a match between running processes and seeing if there's any process that comes from ` $GOPATH/bin ` or not. And maybe let the user type in the PID if not found.

On Unix you could try `` pwdx `pgrep binaryName` `` but I don't think that's a particularly nice approach.

Another is what the Python plugin is doing with Attach to Local Process which generates a list of running processes (currently broken in the latest IDEA 2016.3 EAP, https://youtrack.jetbrains.com/issue/IDEA-161124 )

Hope it helps.

**Comment by hrobertson** [comment]
As it's not possible to run a generic delve server and then specify the package to debug via the API, users would have to start their debug server first, then go to IDEA and run the debug tool to connect to the existing debug session.

Because of this there need to be two distinct methods of debugging - When using localhost the plugin can start the delve server itself, when using a remote delve server it can't.

I've created this feature request for Delve which would facilitate remote debugging with a single click via the IDE: https://github.com/derekparker/delve/issues/648
